Aplib.Core/ExposedQueue.cs

```diff
@@ -0,0 +1,193 @@
+            _array = array;
+            _head = MaxCount - 1;
+            Count = count == -1 ? MaxCount : count;
+        }
+
+        /// <summary>
+        /// Gets the element at the specified index.
+        /// </summary>
+        /// <param name="index">The index of the element to get.</param>
+        /// <returns>The element at the specified index.</returns>
+        public T this[int index]
+        {
+            get => _array[(index + _head + 1) % MaxCount];
+            private set => _array[(index + _head + 1) % MaxCount] = value;
+        }
+
+        /// <summary>
+        /// Decrements the head of the array.
+        /// </summary>
+        private void DecrementHead() => _head = (_head - 1 + MaxCount) % MaxCount;
+
+        /// <summary>
+        /// Puts an element at the start of the array.
+        /// </summary>
+        /// <param name="value">The element to add to the array</param>
+        public void Put(T value)
+        {
+            _array[_head] = value;
+            DecrementHead();
+            Count++;
+        }
+
+        /// <inheritdoc/>
+        public void Add(T item) => Put(item);
+
+        /// <summary>
+        /// Gets the element at the head of the array.
+        /// </summary>
+        /// <returns>The element at the head of the array</returns>
+        public T GetHead() => _array[_head];
+
+        /// <summary>
+        /// Gets the first element of the array.
+        /// </summary>
+        /// <returns>The last element of the array</returns>
+        public T GetFirst() => this[0];
+
+        /// <summary>
+        /// Copies the circular array to an array.
+        /// The head should be the last element of the array.
+        /// Copies from start to end inclusive.
+        /// </summary>
+        /// <param name="array">The array to copy to."</param>
+        /// <param name="arrayIndex">The start index of the range to copy.</param>
+        /// <param name="endIndex">The end index of the range to copy.</param>
+        /// <returns>The circular array as a normal array</returns>
+        public void CopyTo(T[] array, int arrayIndex, int endIndex)
+        {
+            if (arrayIndex < 0 || arrayIndex >= Count)
+                throw new ArgumentOutOfRangeException(nameof(arrayIndex));
+            if (endIndex < 0 || endIndex >= Count)
+                throw new ArgumentOutOfRangeException(nameof(endIndex));
+            if (arrayIndex > endIndex)
+                throw new ArgumentException("Start index must be less than or equal to end index.");
+            if (array.Length < endIndex - arrayIndex + 1)
+                throw new ArgumentException("Array is too small to copy the range.");
+
+            for (int i = 0; i < endIndex - arrayIndex + 1; i++)
+                array[i] = this[arrayIndex + i];
+        }
+
+        /// <inheritdoc/>
+        public void CopyTo(T[] array, int arrayIndex) => CopyTo(array, arrayIndex, arrayIndex + Count - 1);
```

Comment:
Maybe nice to have a default value as well?
```suggestion
        public void CopyTo(T[] array, int startIndex = 0) => CopyTo(array, arrayIndex, arrayIndex + Count - 1);
```

Reply:
Same as other comment, sonar wants it to match the interface, which has no default parameter